src/Clause.ts

```diff
@@ -51,16 +52,85 @@ export default class Clause extends Builder {
       // <emu-clause id=foo aoid> === <emu-clause id=foo aoid=foo>
       this.aoid = node.id;
     }
+
+    let header = this.node.firstElementChild;
+    while (header != null && header.tagName === 'SPAN' && header.children.length === 0) {
+      // skip oldids
+      header = header.nextElementSibling;
+    }
+    if (header == null || header.tagName !== 'H1') {
+      this.spec.warn({
+        type: 'node',
+        ruleId: 'missing-header',
+        message: `clause doesn't have a header, found: ${header?.tagName}`,
+        node: this.node,
+      });
+      header = null;
+    } else {
+      this.buildStructuredHeader(header);
+    }
+    this.header = header;
   }
 
-  buildHeader() {
-    if (!this.number) return;
+  buildStructuredHeader(header: Element) {
+    const dl = header.nextElementSibling;
+    if (dl == null || dl.tagName !== 'DL' || !dl.classList.contains('header')) {
+      return;
+    }
+    // if we find such a DL, treat this as a structured header
+
+    const type = this.node.getAttribute('type') ?? 'unknown';
+    this.node.removeAttribute('type'); // TODO maybe leave it in; this is just to minimize the diff
+
+    const { name, formattedHeader, formattedParams } = parseStructuredHeaderH1(this.spec, header);
+    if (type === 'numeric method' && name != null && !name.includes('::')) {
+      this.spec.warn({
+        type: 'contents',
+        ruleId: 'numberic-method-for',
+        message: `numeric methods should be of the form \`Type::operation\``,
+        node: dl,
+        nodeRelativeLine: 1,
+        nodeRelativeColumn: 1,
+      });
+    }
+    if (formattedHeader != null) {
+      header.innerHTML = formattedHeader;
+    }
 
-    const numElem = this.spec.doc.createElement('span');
-    numElem.setAttribute('class', 'secnum');
-    numElem.textContent = this.number;
-    this.header.insertBefore(this.spec.doc.createTextNode(' '), this.header.firstChild);
-    this.header.insertBefore(numElem, this.header.firstChild);
+    const { description, for: _for } = parseStructuredHeaderDl(this.spec, type, dl);
+
+    const paras = formatPreamble(
+      this.spec,
+      this.node,
+      dl,
+      type,
+      name ?? 'UNKNOWN',
+      formattedParams ?? 'UNPARSEABLE ARGUMENTS',
```

Comment:
If these are null, why not just fail?

Reply:
Mostly ecmarkup tries hard to produce some output unless you pass `--strict`. That makes it easier to iterate on the stuff you are actually trying to work on and fix issues in whatever order you want.

Reply:
Well surely we should output a warning in this case, right?

Reply:
[We do](https://github.com/tc39/ecmarkup/blob/7837fc12c4741637e44d7b892e4a7466b46db4dc/src/header-parser.ts#L28), at the point at which parsing fails. The linked line is the only path where `parseStructuredHeaderH1` can return `null`. (I can tweak the structure of that method to make that more obvious, I suppose.)

Reply:
Made the above tweak.